uri/BaseUri.php

```diff
@@ -269,6 +273,41 @@ public function hasIdn(): bool
         return Converter::isIdn($this->uri->getHost());
     }
 
+    /**
+     * Redact the password component and any query pairs whose key is submitted.
+     *
+     * All redacted values will be replaced by the 5-star mask.
+     * The return value MAY not be a valid URI
+     *
+     * @param string ...$names
+     */
+    public function redactSensitiveParameters(string ...$names): string
+    {
+        $components = UriString::parse($this);
+        if ($components['pass'] !== null) {
+            $components['pass'] = self::COMPONENT_MASK;
+        }
+
+        $currentQuery = $components['query'];
+        if ([] === $names || null === $currentQuery || !str_contains($currentQuery, '=')) {
+            return UriString::build($components);
+        }
+
+        $names = array_map(Encoder::decodeAll(...), $names);
+        $pairs = [];
+        foreach (explode('&', $currentQuery) as $part) {
```

Comment:
Doesn't this duplicate QueryString parsing?

Reply:
No it does not it just bypassed most of the validation hence why it returns a string representation that may not represent a valid URI.